File: 찾아라_프로그래밍_마에스터/게임_맵_최단거리/solve.cpp
#include<vector>
#include<iostream>
#include<algorithm>
#include<queue>

using namespace std;

int solution(vector<vector<int> > maps)
{
    int answer = 0;
    
    int col = maps.size();
    int row = maps[0].size();
    
    int visited[101][101] = {0,};
    queue<pair<int,int>> que;
    
    int dx[] = {0,0,-1,1};
    int dy[] = {-1,1,0,0};
    
    visited[0][0] = 1;
    que.push(make_pair(0,0));

    while(!que.empty()){
        queue<pair<int,int>> que2;

        while(!que.empty()){
            int x = que.front().first;
            int y = que.front().second;
            int cnt = visited[x][y];
            que.pop();
            
            for(int i=0; i<4; i++){
                int nx = x + dx[i];
                int ny = y + dy[i];
                
                if(nx<0 || ny<0 || nx>=col || ny>=row) continue;
                if(visited[nx][ny] || !maps[nx][ny]) continue;
                
                visited[nx][ny] = cnt+1;
                que2.push(make_pair(nx,ny));
            }
        }
        que = que2;
    }
    
    visited[col-1][row-1] ? answer = visited[col-1][row-1] : answer = -1;
    return answer;
}

File: 그래프/가장_먼_노드/solve.py
from queue import Queue

visited = []
eg = []

def bfs(edge):
    global visited, eg
    start = 1
    
    que = Queue()
    que.put(start)
    visited[start] = True

    while not que.empty():
        que2 = Queue()
        cnt = que.qsize()

        while not que.empty():
            cur = que.get()

            for e in eg[cur]:
                if not visited[e]:
                    que2.put(e)
                    visited[e] = True
        
        que = que2
        
    return cnt

def solution(n, edge):
    global visited, eg
    
    visited = [False for i in range(n+1)]
    eg = [[] for i in range(n+1)]
    
    for i in range(len(edge)):
        f = edge[i][0]
        t = edge[i][1]
        eg[f].append(t)
        eg[t].append(f)
    
    return bfs(edge)
 
